File: plsql-frontend/src/main/java/org/sonar/plsqlopen/matchers/MethodMatcher.java
/*
 * Sonar PL/SQL Plugin (Community)
 * Copyright (C) 2015-2017 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plsqlopen.matchers;

import static java.util.stream.Collectors.toList;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import org.sonar.plugins.plsqlopen.api.PlSqlGrammar;

import com.google.common.base.Preconditions;
import com.google.common.collect.Lists;
import com.sonar.sslr.api.AstNode;
import com.sonar.sslr.api.AstNodeType;

public class MethodMatcher {

    private static final AstNodeType[] VARIABLE_OR_IDENTIFIER = { PlSqlGrammar.VARIABLE_NAME, PlSqlGrammar.IDENTIFIER_NAME };
    
    private NameCriteria methodNameCriteria;
    private NameCriteria packageNameCriteria;
    private NameCriteria schemaNameCriteria;
    private int parameterCount;
    private boolean shouldCheckParameters = true;
    private boolean schemaIsOptional = false;
    private String methodName;

    private MethodMatcher() {
        // instances should be created using the create method
    }
    
    public String getMethodName() {
        return methodName;
    }
    
    public static MethodMatcher create() {
        return new MethodMatcher();
    }
    
    public MethodMatcher name(String methodNameCriteria) {
        return name(NameCriteria.is(methodNameCriteria));
    }
    
    public MethodMatcher name(NameCriteria methodNameCriteria) {
        Preconditions.checkState(this.methodNameCriteria == null);
        this.methodNameCriteria = methodNameCriteria;
        return this;
    }
    
    public MethodMatcher packageName(String packageNameCriteria) {
        return packageName(NameCriteria.is(packageNameCriteria));
    }
    
    public MethodMatcher packageName(NameCriteria packageNameCriteria) {
        Preconditions.checkState(this.packageNameCriteria == null);
        this.packageNameCriteria = packageNameCriteria;
        return this;
    }
    
    public MethodMatcher schema(String schemaNameCriteria) {
        return schema(NameCriteria.is(schemaNameCriteria));
    }
    
    public MethodMatcher schema(NameCriteria schemaNameCriteria) {
        Preconditions.checkState(this.schemaNameCriteria == null);
        this.schemaNameCriteria = schemaNameCriteria;
        return this;
    }
    
    public MethodMatcher withNoParameterConstraint() {
        Preconditions.checkState(this.parameterCount == 0);
        this.shouldCheckParameters = false;
        return this;
    }

    public MethodMatcher schemaIsOptional() {
        this.schemaIsOptional = true;
        return this;
    }
    
    public MethodMatcher addParameter() {
        Preconditions.checkState(this.shouldCheckParameters);
        this.parameterCount++;
        return this;
    }
    
    public MethodMatcher addParameters(int quantity) {
        Preconditions.checkState(this.shouldCheckParameters);
        this.parameterCount += quantity;
        return this;
    }
    
    public List<AstNode> getArguments(AstNode node) {
        AstNode arguments = node.getFirstChild(PlSqlGrammar.ARGUMENTS);
        if (arguments != null) {
            return arguments.getChildren(PlSqlGrammar.ARGUMENT);
        }
        
        return new ArrayList<>();
    }
    
    public List<AstNode> getArgumentsValues(AstNode node) {
        return getArguments(node).stream().map(AstNode::getLastChild).collect(toList());
    }

    public boolean matches(AstNode originalNode) {
        AstNode node = normalize(originalNode);
        LinkedList<AstNode> nodes = Lists.newLinkedList(node.getChildren(VARIABLE_OR_IDENTIFIER));
        
        if (nodes.isEmpty()) {
            return false;
        }
        
        boolean matches = true;
        
        matches &= nameAcceptable(nodes.removeLast(), methodNameCriteria);
        
        if (packageNameCriteria != null) {
            matches &= !nodes.isEmpty() && nameAcceptable(nodes.removeLast(), packageNameCriteria);
        }
        
        if (schemaNameCriteria != null) {
            matches &= (schemaIsOptional && nodes.isEmpty()) || 
                    (!nodes.isEmpty() && nameAcceptable(nodes.removeLast(), schemaNameCriteria));
        }
        
        return matches && nodes.isEmpty() && argumentsAcceptable(originalNode);
    }

    private boolean nameAcceptable(AstNode node, NameCriteria criteria) {
        methodName = node.getTokenOriginalValue();
        return criteria.matches(methodName);
    }

    private boolean argumentsAcceptable(AstNode node) {
        return !shouldCheckParameters || getArguments(node).size() == parameterCount;
    }
    
    private static AstNode normalize(AstNode node) {
        if (node.getType() == PlSqlGrammar.METHOD_CALL || node.getType() == PlSqlGrammar.CALL_STATEMENT) {
            AstNode child = normalize(node.getFirstChild());
            if (child.getFirstChild().getType() == PlSqlGrammar.HOST_AND_INDICATOR_VARIABLE) {
                child = child.getFirstChild();
            }
            return child;
        }
        return node;
    }
    
}

File: zpa-core/src/main/kotlin/org/sonar/plugins/plsqlopen/api/matchers/MethodMatcher.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2021 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plugins.plsqlopen.api.matchers

import com.sonar.sslr.api.AstNode
import org.sonar.plugins.plsqlopen.api.PlSqlGrammar
import org.sonar.plugins.plsqlopen.api.squid.SemanticAstNode
import org.sonar.plugins.plsqlopen.api.symbols.PlSqlType

class MethodMatcher private constructor()
{

    private var methodNameCriteria: NameCriteria? = null
    private var packageNameCriteria: NameCriteria? = null
    private var schemaNameCriteria: NameCriteria? = null
    private var shouldCheckParameters = true
    private var schemaIsOptional = false
    var methodName: String = ""
        private set
    private val expectedArgumentTypes = ArrayList<PlSqlType>()

    fun name(methodNameCriteria: String) =
        name(NameCriteria.`is`(methodNameCriteria))

    fun name(methodNameCriteria: NameCriteria) = apply {
        check(this.methodNameCriteria == null)
        this.methodNameCriteria = methodNameCriteria
    }

    fun packageName(packageNameCriteria: String) =
        packageName(NameCriteria.`is`(packageNameCriteria))

    fun packageName(packageNameCriteria: NameCriteria) = apply {
        check(this.packageNameCriteria == null)
        this.packageNameCriteria = packageNameCriteria
    }

    fun schema(schemaNameCriteria: String) =
        schema(NameCriteria.`is`(schemaNameCriteria))

    fun schema(schemaNameCriteria: NameCriteria) = apply {
        check(this.schemaNameCriteria == null)
        this.schemaNameCriteria = schemaNameCriteria
    }

    fun withNoParameterConstraint() = apply {
        check(this.expectedArgumentTypes.isEmpty())
        this.shouldCheckParameters = false
    }

    fun schemaIsOptional() = apply {
        this.schemaIsOptional = true
    }

    fun addParameter() =
        addParameter(PlSqlType.UNKNOWN)

    fun addParameter(type: PlSqlType) = apply {
        check(this.shouldCheckParameters)
        expectedArgumentTypes.add(type)
    }

    fun addParameters(quantity: Int) = apply {
        for (i in 0 until quantity) {
            addParameter(PlSqlType.UNKNOWN)
        }
    }

    fun addParameters(vararg types: PlSqlType) = apply {
        check(this.shouldCheckParameters)
        for (type in types) {
            addParameter(type)
        }
    }

    fun getArguments(node: AstNode): List<AstNode> {
        val arguments = node.getFirstChild(PlSqlGrammar.ARGUMENTS)
        return if (arguments != null) {
            arguments.getChildren(PlSqlGrammar.ARGUMENT)
        } else ArrayList()

    }

    fun getArgumentsValues(node: AstNode) =
        getArguments(node).map { it.lastChild }.toList()

    fun matches(originalNode: AstNode): Boolean {
        val node = normalize(originalNode)
        val nodes = node.getChildren(PlSqlGrammar.VARIABLE_NAME, PlSqlGrammar.IDENTIFIER_NAME)

        if (nodes.isEmpty()) {
            return false
        }

        var matches =  methodNameCriteria?.let { nameAcceptable(nodes.removeAt(nodes.lastIndex), it) } ?: true

        packageNameCriteria?.let {
            matches = matches and (nodes.isNotEmpty() && nameAcceptable(nodes.removeAt(nodes.lastIndex), it))
        }

        schemaNameCriteria?.let {
            matches = matches and (schemaIsOptional && nodes.isEmpty() || nodes.isNotEmpty() && nameAcceptable(nodes.removeAt(nodes.lastIndex), it))
        }

        return matches && nodes.isEmpty() && argumentsAcceptable(originalNode)
    }

    private fun nameAcceptable(node: AstNode, criteria: NameCriteria): Boolean {
        methodName = node.tokenOriginalValue
        return criteria.matches(methodName)
    }

    private fun argumentsAcceptable(node: AstNode): Boolean {
        val arguments = getArguments(node)
        return !shouldCheckParameters || arguments.size == expectedArgumentTypes.size && argumentTypesAreCorrect(arguments)
    }

    private fun argumentTypesAreCorrect(arguments: List<AstNode>): Boolean {
        var result = true
        for ((i, type) in expectedArgumentTypes.withIndex()) {
            val actualArgument = arguments[i].firstChild
            result = result and (type === PlSqlType.UNKNOWN || type === semantic(actualArgument).plSqlType)
        }
        return result
    }

    private fun normalize(node: AstNode): AstNode {
        if (node.type === PlSqlGrammar.METHOD_CALL || node.type === PlSqlGrammar.CALL_STATEMENT) {
            var child = normalize(node.firstChild)
            if (child.firstChild.type === PlSqlGrammar.HOST_AND_INDICATOR_VARIABLE) {
                child = child.firstChild
            }
            return child
        }
        return node
    }

    companion object {
        @JvmStatic
        fun create(): MethodMatcher {
            return MethodMatcher()
        }

        fun semantic(node: AstNode): SemanticAstNode {
            return node as SemanticAstNode
        }
    }

}
